File: android/slack/src/main/java/com/prince/slack/SlackIntegration.java
package com.prince.slack;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.view.inputmethod.EditorInfo;

import androidx.annotation.Nullable;

import com.prince.kbd.core.CommandSpec;
import com.prince.kbd.core.IntegrationContext;
import com.prince.kbd.core.IntegrationSession;
import com.prince.kbd.core.KeyValueStore;
import com.prince.kbd.core.KeyboardIntegration;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Slack integration. Contributes {@code /slack} (and {@code /msg} alias): posts the
 * message verbatim to the user's default channel via {@code chat.postMessage}. A leading
 * {@code #channel} or {@code <#CID>} token overrides the channel.
 */
public final class SlackIntegration implements KeyboardIntegration {

    private static final Set<String> AFFINITY = new HashSet<>(Arrays.asList(
            "com.notion.id",
            "com.google.android.gm",
            "com.android.chrome",
            "com.linkedin.android",
            "org.telegram.messenger",
            "com.whatsapp"));

    @Override public String id() { return "slack"; }

    @Override
    @Nullable
    public IntegrationSession activate(EditorInfo info, IntegrationContext ctx) {
        return null;
    }

    @Override
    public List<CommandSpec> commands() {
        return Arrays.asList(
                new CommandSpec("slack", "Slack", "💬", true, this::handleSlack, AFFINITY),
                new CommandSpec("msg",   "Slack message", "💬", true, this::handleSlack, AFFINITY));
    }

    private void handleSlack(String prompt, IntegrationContext ctx) {
        KeyValueStore store = ctx.store("slack");
        SlackAuth auth = new SlackAuth(store);
        if (!auth.isSignedIn()) {
            ctx.showBanner("Connect Slack in the Turtle app", 1800L);
            ctx.openScreen("slack-connect");
            return;
        }

        Resolved r = resolveChannel(prompt, store);
        if (r.channelId == null || r.channelId.isEmpty()) {
            ctx.showBanner("Pick a Slack channel in the Turtle app", 1800L);
            ctx.openScreen("slack-connect");
            return;
        }
        if (r.text == null || r.text.trim().isEmpty()) {
            ctx.showBanner("Type a message after /slack", 1500L);
            return;
        }

        ctx.showBanner("💬 Sending to #" + r.channelName + "…", 1000L);

        final Context appContext = ctx.appContext();
        final String token = auth.accessToken();
        final String userPrompt = r.text;
        final String channelName = r.channelName;

        new SlackClient(token).postMessage(r.channelId, r.text, new SlackClient.PostCallback() {
            @Override public void onSuccess(String channelId, String ts, String permalink) {
                post(() -> SlackResultNotifier.notifySuccess(appContext, channelName, permalink));
            }
            @Override public void onError(String reason) {
                post(() -> SlackResultNotifier.notifyError(appContext, userPrompt, reason));
            }
        });
    }

    private static final class Resolved {
        final String channelId;
        final String channelName;
        final String text;
        Resolved(String id, String name, String text) {
            this.channelId = id; this.channelName = name; this.text = text;
        }
    }

    /** Strip a leading {@code #name} or {@code <#CID>} channel mention from {@code prompt}. */
    private static Resolved resolveChannel(String prompt, KeyValueStore store) {
        String defaultId = store.getString(SlackKeys.DEFAULT_CHANNEL, "");
        String defaultName = store.getString(SlackKeys.DEFAULT_CHANNEL_NAME, "");
        if (prompt == null) return new Resolved(defaultId, defaultName, "");
        String trimmed = prompt.trim();
        // <#CID|name> form — Slack's UI inserts this when you pick a channel
        if (trimmed.startsWith("<#")) {
            int gt = trimmed.indexOf('>');
            if (gt > 2) {
                String inside = trimmed.substring(2, gt);
                int pipe = inside.indexOf('|');
                String cid = pipe >= 0 ? inside.substring(0, pipe) : inside;
                String name = pipe >= 0 ? inside.substring(pipe + 1) : cid;
                String body = trimmed.substring(gt + 1).trim();
                return new Resolved(cid, name, body);
            }
        }
        if (trimmed.startsWith("#")) {
            int sp = indexOfWhitespace(trimmed);
            String name = (sp > 0 ? trimmed.substring(1, sp) : trimmed.substring(1)).toLowerCase();
            String body = sp > 0 ? trimmed.substring(sp + 1).trim() : "";
            String id = lookupChannelId(store, name);
            if (id != null) return new Resolved(id, name, body);
            // Unknown channel — fall back to default, keeping the typo visible to the user.
        }
        return new Resolved(defaultId, defaultName, trimmed);
    }

    private static int indexOfWhitespace(String s) {
        for (int i = 0; i < s.length(); i++) {
            if (Character.isWhitespace(s.charAt(i))) return i;
        }
        return -1;
    }

    /** @return channel id for {@code name} from the cached map, or null if not present. */
    @Nullable
    private static String lookupChannelId(KeyValueStore store, String name) {
        String id = store.getString("channel_map." + name, "");
        return id == null || id.isEmpty() ? null : id;
    }

    private static void post(Runnable r) {
        new Handler(Looper.getMainLooper()).post(r);
    }
}

File: ios/TurtleKeyboardExtension/Integration/Slack/SlackIntegration.swift
import Foundation
#if os(iOS)
import UIKit

/// Commands-only Slack integration. Contributes `/slack` and `/msg` —
/// posts the user's message to their default channel (or a channel they
/// override with a leading `#name`). Mirrors Android's `SlackIntegration`.
///
/// Behavior is fire-and-forget:
///   1. User taps Send on `/slack <message>`.
///   2. Banner flashes "Sending to #channel…".
///   3. Background: `chat.postMessage`.
///   4. Banner shows the result. On success the permalink is copied to the
///      clipboard so the user can paste it anywhere.
final class SlackIntegration: KeyboardIntegration {

    let id = "slack"

    func commands() -> [CommandSpec] {
        [
            CommandSpec(
                name: "slack", label: "Slack", emoji: "💬", needsPrompt: true,
                handler: { prompt, ctx in Self.handle(prompt: prompt, ctx: ctx) }
            ),
            CommandSpec(
                name: "msg", label: "Slack message", emoji: "💬", needsPrompt: true,
                handler: { prompt, ctx in Self.handle(prompt: prompt, ctx: ctx) }
            ),
        ]
    }

    static func handle(prompt: String, ctx: IntegrationContext) {
        let store = ctx.store
        let token = store.string(forKey: SlackKeys.accessToken, fallback: "")
        guard !token.isEmpty else {
            ctx.showBanner("Connect Slack in the Turtle app", autoHideMs: 1800)
            ctx.openScreen("slack-connect")
            return
        }

        let resolved = resolveChannel(prompt: prompt, store: store)
        guard !resolved.channelId.isEmpty else {
            ctx.showBanner("Pick a Slack channel in the Turtle app", autoHideMs: 1800)
            ctx.openScreen("slack-connect")
            return
        }
        let body = resolved.text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !body.isEmpty else {
            ctx.showBanner("Type a message after /slack", autoHideMs: 1500)
            return
        }

        ctx.showBusy("💬 Sending to #\(resolved.channelName)…")

        Task {
            do {
                let result = try await SlackClient.postMessage(
                    accessToken: token,
                    channelId: resolved.channelId,
                    text: body)
                await MainActor.run {
                    if let link = result.permalink, !link.isEmpty {
                        UIPasteboard.general.string = link
                        ctx.showBanner("💬 Posted to #\(resolved.channelName) — link copied",
                                       autoHideMs: 2000)
                    } else {
                        ctx.showBanner("💬 Posted to #\(resolved.channelName)",
                                       autoHideMs: 1500)
                    }
                }
            } catch {
                await MainActor.run {
                    ctx.showBanner("⚠️ Slack: \(error.localizedDescription)", autoHideMs: 2500)
                }
            }
        }
    }

    // MARK: - Channel resolution

    private struct Resolved {
        let channelId: String
        let channelName: String
        let text: String
    }

    /// Pull a leading `#name` or `<#CID|name>` token off `prompt`. If the
    /// name matches a known channel (cached at connect time), route there.
    /// Otherwise use the stored default.
    private static func resolveChannel(prompt: String, store: SplitStore) -> Resolved {
        let defaultId = store.string(forKey: SlackKeys.defaultChannel, fallback: "")
        let defaultName = store.string(forKey: SlackKeys.defaultChannelName, fallback: "")
        let trimmed = prompt.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            return Resolved(channelId: defaultId, channelName: defaultName, text: "")
        }
        // <#CID|name> form (used when Slack inserts a channel mention)
        if trimmed.hasPrefix("<#"), let close = trimmed.firstIndex(of: ">") {
            let inside = String(trimmed[trimmed.index(trimmed.startIndex, offsetBy: 2)..<close])
            let cid: String
            let name: String
            if let pipe = inside.firstIndex(of: "|") {
                cid = String(inside[inside.startIndex..<pipe])
                name = String(inside[inside.index(after: pipe)...])
            } else {
                cid = inside
                name = inside
            }
            let body = String(trimmed[trimmed.index(after: close)...])
                .trimmingCharacters(in: .whitespacesAndNewlines)
            return Resolved(channelId: cid, channelName: name, text: body)
        }
        // #channel-name form — only honored when the channel id is cached.
        if trimmed.hasPrefix("#") {
            let afterHash = trimmed.dropFirst()
            let firstSpace = afterHash.firstIndex(where: { $0.isWhitespace })
            let nameSlice = firstSpace.map { afterHash[afterHash.startIndex..<$0] } ?? afterHash
            let name = String(nameSlice).lowercased()
            let body = firstSpace.map {
                String(afterHash[afterHash.index(after: $0)...])
                    .trimmingCharacters(in: .whitespacesAndNewlines)
            } ?? ""
            let id = store.string(forKey: SlackKeys.channelMapPrefix + name, fallback: "")
            if !id.isEmpty {
                return Resolved(channelId: id, channelName: name, text: body)
            }
            // Unknown channel — fall through to default with prefix kept.
        }
        return Resolved(channelId: defaultId, channelName: defaultName, text: trimmed)
    }
}
#endif
